modules/mod-deep-learning/ModelCard.h

```diff
@@ -0,0 +1,212 @@
+
+   ModelCard.h
+   Hugo Flores Garcia
+
+******************************************************************/
+/**
+
+\class ModelCard
+\brief model metadata for deep learning models
+
+*/
+/*******************************************************************/
+
+#pragma once
+
+#include <torch/script.h>
+#include <torch/torch.h>
+
+#include <rapidjson/document.h>
+#include <rapidjson/schema.h>
+#include "rapidjson/prettywriter.h" 
+#include <rapidjson/writer.h>
+#include <wx/string.h>
+#include <wx/log.h>
+
+#include "MemoryX.h"
+#include "AudacityException.h"
+
+using Doc = rapidjson::Document;
+
+// this exception should be caught internally, but we 
+// derive from MessageBoxException just in case it needs to 
+// get handled by Audacity
+class InvalidModelCardDocument final : public MessageBoxException
+{
+public:
+   InvalidModelCardDocument(const TranslatableString msg, std::string trace): 
+                           m_msg(msg), 
+                           m_trace(std::move(trace)),
+                            MessageBoxException {
+                               ExceptionType::Internal,
+                               XO("Invalid Model Card Document")
+                            }
+   { 
+   }
+
+   // detailed internal error message
+   virtual const char* what() const throw () 
+   {
+      wxLogError(m_msg.Translation());
+      if (!m_trace.empty()) 
+         wxLogError(wxString(m_trace)); 
+      // TODO: also print the document
+      // TODO: check if document is nullptr
+      return m_msg.Translation().c_str();
+   }
+
+   // user facing message
+   virtual TranslatableString ErrorMessage() const
+      { return XO("Model Card Error: %s")
+               .Format(m_msg);}
+
+   const TranslatableString m_msg;
+   const std::string m_trace;
+};
+
+namespace parsers 
+{
+   Doc ParseString(const std::string &json);
+   Doc ParseFile(const std::string &path);
+}
+
+// validators that accept a default value are no-throw
+namespace validators
+{
+
+   std::string tryGetString(const std::string &key, const Doc& doc);
+   std::string tryGetString(const std::string &key, const Doc& doc, const std::string &defaultValue);
+
+   int tryGetInt(const std::string &key, const Doc& doc);
+   int tryGetInt(const std::string &key, const Doc& doc, int defaultValue);
+
+   std::vector<std::string> tryGetStringArray(const std::string &key, const Doc& doc, bool lower=false);
+                                             
+   uint64_t tryGetUint64(const std::string &key, const Doc& doc);
+   uint64_t tryGetUint64(const std::string &key, const Doc& doc, uint64_t defaultValue);
+
+}
+
+class DeepModelManager;
+
+class ModelCard final
+{
+public: 
+   ModelCard() = default;
+   ModelCard(const ModelCard&) = default;
+
+   // returns {author}/{name}
+   std::string GetRepoID() const;
+
+   bool IsSame(const ModelCard& other) const { return (*this) == other; };
+
+private:
+   friend class DeepModelManager;
```

Comment:
"long-distance" friendship, with a class defined in a different header file, is something to avoid.  It implies I can't really understand ModelCard without understanding DeepModelManager.

Why not just publicize more methods of ModelCard?

What are invariant conditions of the ModelCard object that must be maintained by functions in its public interface?  If you can state them, and if use of the mutating methods that are made public when you remove the friend declaration don't lead to violations of them -- then DeepModelManager doesn't really need to be a friend.

Reply:
d19180bd489b299b7e7a8caee2c5a9f13bd2e689